library_generation/utils/generation_config_comparator.py

```diff
@@ -0,0 +1,379 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from collections import defaultdict
+from enum import Enum
+from typing import Dict
+from typing import List
+
+from library_generation.model.gapic_config import GapicConfig
+from library_generation.model.generation_config import GenerationConfig
+from library_generation.model.library_config import LibraryConfig
+from library_generation.utilities import get_library_name
+
+
+class ChangeType(Enum):
+    GOOGLEAPIS_COMMIT = 1
+    REPO_LEVEL_CHANGE = 2
+    LIBRARIES_ADDITION = 3
+    LIBRARIES_REMOVAL = 4
+    LIBRARY_LEVEL_CHANGE = 5
+    GAPIC_ADDITION = 6
+    GAPIC_REMOVAL = 7
+
+
+class HashLibrary:
+    """
+    Data class to group a LibraryConfig object and its hash value together.
+    """
+
+    def __init__(self, hash_value: int, library: LibraryConfig):
+        self.hash_value = hash_value
+        self.library = library
+
+
+class ConfigChange:
+    def __init__(self, changed_param: str, latest_value: str, library_name: str = ""):
+        self.changed_param = changed_param
+        self.latest_value = latest_value
+        self.library_name = library_name
+
+
+def compare_config(
+    baseline_config: GenerationConfig, latest_config: GenerationConfig
+) -> Dict[ChangeType, list[ConfigChange]]:
+    """
+    Compare two GenerationConfig object and output a mapping from ConfigChange
+    to a list of ConfigChange objects.
+    All libraries in the latest configuration will be affected if one of the
+    repository level parameters is changed.
+
+    :param baseline_config: the baseline GenerationConfig object
+    :param latest_config: the latest GenerationConfig object
+    :return: a mapping from ConfigChange to a list of ConfigChange objects.
+    """
+    diff = defaultdict(list[ConfigChange])
+    if baseline_config.googleapis_commitish != latest_config.googleapis_commitish:
+        diff[ChangeType.GOOGLEAPIS_COMMIT] = []
+    if baseline_config.gapic_generator_version != latest_config.gapic_generator_version:
+        config_change = ConfigChange(
```

Comment:
Can we iterate over all fields in `GenerationConfig`? Same thing for the `LibraryConfig` below. It could be a separate PR.

Reply:
> Can we iterate over all fields in GenerationConfig? Same thing for the LibraryConfig below.

Does the goal is to reduce the amount of code?

Reply:
Yes, the logic is duplicated for every field.

Reply:
To iterate field of a class, I think we need a enum type or a list of strings to specify the field?

Reply:
Does something like [this](https://stackoverflow.com/questions/11637293/iterate-over-object-attributes-in-python) help?

Reply:
I don't know python can do this, thanks for the help. I'll do some refactor then.